1117 jump to main content

skspade opened the pull request:
# Describe 
- Adds ability to press the "Tab" key to reveal a hidden "Jump to conent" link. Clicking or hitting enter will bring the post listings regardless of page into focus. 
- Prevents `validURL` from throwing exception into the browser console.

# Video 

https://github.com/LemmyNet/lemmy-ui/assets/43147936/203ce936-8070-4ea3-8100-0be18233f3ad


# Issue 
closes #1117

SleeplessOne1917 commented on src/assets/css/main.css:
```diff
@@ -430,3 +430,19 @@ br.big {
+  position: absolute;
+  top: -40px;
+  left: 0;
+  background: #000;
```
Could you please use bootstrap class names for styling where possible? You could give the skip link the following classes: "bg-dark text-light p-2 text-decoration-none position-absolute start-0 z-3" The only properties you'll still need defined on this css class are top and transition.

skspade replied:
The bootstrap color classes are confusing me a bit. For example `bg-dark` is giving me a light color background. `text-light` is giving me dark text. The opposite is true too. Look at my screenshots below.
![Screenshot from 2023-06-11 13-11-04](https://github.com/LemmyNet/lemmy-ui/assets/43147936/4b065726-39e3-43bd-b2c9-57ef8843c5e5)
![Screenshot from 2023-06-11 13-11-29](https://github.com/LemmyNet/lemmy-ui/assets/43147936/d70c3a24-0663-429c-b6ba-81818438df6f)

SleeplessOne1917 commented on src/server/index.tsx:
```diff
@@ -393,6 +393,13 @@ async function createSsrHtml(root: string, isoData: IsoDataOptionalSite) {
+    <a
+    href="#main-content"
+    class="skip-link"
+    onclick="document.getElementById('main-content').focus(); return false;"
```
You don't have to make this function a string. You could make it:

```tsx
<a onClick={() => {
    document.getElementById('main-content')?.focus(); 
}}
>
    Jump to content
</a>
```

skspade replied:
Are you able to use JSX syntax here? This is in the `createSsrHtml` function which does not seem to be JSX but a HTML in a template string.

SleeplessOne1917 replied:
That is a good point. I didn't realize this wasn't JSX in the small window github gave me.

SleeplessOne1917 commented on src/server/index.tsx:
```diff
@@ -393,6 +393,13 @@ async function createSsrHtml(root: string, isoData: IsoDataOptionalSite) {
+    class="skip-link"
+    onclick="document.getElementById('main-content').focus(); return false;"
+    >
+      Jump to content
```
This is going to need a translation. [Please add one to the translations repo](https://github.com/LemmyNet/lemmy-translations).

SleeplessOne1917 reviewed (changes requested):
Please address the changes I requested in the other comments. In addition to that, this jump to content link should be able to work in more places than just the post listings. As it is now, this will only allow users to jump to content on their home screen or a community screen.

skspade commented:
Great! I'll get to work on your suggestions tonight. Thanks for the review

skspade commented:
> This is going to need a translation. [Please add one to the translations repo](https://github.com/LemmyNet/lemmy-translations). 

https://github.com/LemmyNet/lemmy-translations/pull/50

skspade commented:
I re-implemented the feature, this time I wrapped all routes at the entry point. I feel like this is effectively skips the navigation elements on every page and in the future other developers do not have to worry about wrapping new pages with an element containing the attribute `main-content`

dessalines reviewed (commented):
Externalize that one to a handler function. The convention is to have any onClick actions use linkEvent with a function.

skspade commented:
Sounds good, I can move it into a JSX portion of the code. I put it here since it needs to be the first focusable element on any page. But I agree consistency is important. I'll get to work on this in the evening

skspade commented:
Updated Video showing the functionality is the same when written in JSX
https://github.com/LemmyNet/lemmy-ui/assets/43147936/62b5f0eb-6a22-49f6-bd3c-30e718bf21f4

skspade commented:
https://github.com/LemmyNet/lemmy-ui/assets/43147936/f0991bfa-63d5-4ed7-9f1a-c648e525f2f2

skspade commented:
I see that there is conflicts now, I'll fix those tonight

SleeplessOne1917 commented:
@skspade The only conflict is the translations submodule. Try running `git submodule update --remote --recursive` and committing that.

ayan4m1 commented on src/shared/utils.ts:
```diff
@@ -329,7 +329,12 @@ export function isVideo(url: string) {
+  try {
+    new URL(str);
+    return true;
+  } catch (_) {
```
The discard parameter can be omitted entirely here

```jsx
try {
  new URL(str);
  return true;
} catch {
  return false;
}
```

skspade replied:
Good point! I'll make that change too

skspade commented on src/assets/css/main.css:
```diff
@@ -434,3 +434,12 @@ br.big {
+
+.skip-link {
+  top: -40px;
+  transition: top 0.3s ease;
```
True! Since we do not have a user preference for this, I think Its best to just remove the transition.

ayan4m1 replied:
Oh, I meant using a media query: https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion

skspade replied:
Hey thanks for that! Never knew you could query for that! 
I took up this PR to learn more about accessibility so thanks for spreading some knowledge

skspade commented:
@SleeplessOne1917  I believe I got it to a good place. Let me know if you would like me to squash my commits before we merge (if you approve)

SleeplessOne1917 commented:
@skspade  Squashing is fine.

skspade commented:
@SleeplessOne1917 Thanks for the approval! Looks like woodpecker is stalling out after I squashed.  Rebased commit: ` 36033b7` passed, then I squashed to ` d3e1812 `  which doesn't seem to be detected.
